golang/src/user-sessions/main.go

```diff
@@ -0,0 +1,393 @@
+package main
+
+import (
+	"configurate"
+	"database/sql"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"logcabin"
+	"net/http"
+	"os"
+	"queries"
+	"strings"
+
+	"github.com/gorilla/mux"
+	_ "github.com/lib/pq"
+)
+
+var (
+	version = flag.Bool("version", false, "Print the version information")
+	cfgPath = flag.String("config", "/etc/iplant/de/jobservices.yml", "The path to the config file")
+	port    = flag.String("port", "60000", "The port number to listen on")
+	gitref  string
+	appver  string
+	builtby string
+)
+
+func init() {
+	flag.Parse()
+}
+
+// AppVersion prints the version information to stdout
+func AppVersion() {
+	if appver != "" {
+		fmt.Printf("App-Version: %s\n", appver)
+	}
+	if gitref != "" {
+		fmt.Printf("Git-Ref: %s\n", gitref)
+	}
+	if builtby != "" {
+		fmt.Printf("Built-By: %s\n", builtby)
+	}
+}
+
+// App defines the interface for a user-sessions application
+type App interface {
+	Greeting(http.ResponseWriter, *http.Request)
+	GetRequest(http.ResponseWriter, *http.Request)
+	PutRequest(http.ResponseWriter, *http.Request)
+	PostRequest(http.ResponseWriter, *http.Request)
+	DeleteRequest(http.ResponseWriter, *http.Request)
+}
+
+// UserSessionRecord represents a user session stored in the database
+type UserSessionRecord struct {
+	ID      string
+	Session string
+	UserID  string
+}
+
+// convert makes sure that the JSON has the correct format. "wrap" tells convert
+// whether to wrap the object in a map with "session" as the key.
+func convert(record *UserSessionRecord, wrap bool) (map[string]interface{}, error) {
+	var values map[string]interface{}
+	if record.Session != "" {
+		if err := json.Unmarshal([]byte(record.Session), &values); err != nil {
+			return nil, err
+		}
+	}
+	// We don't want the return value wrapped in a session object, so unwrap it
+	// if it is wrapped.
+	if !wrap {
+		if _, ok := values["session"]; ok {
+			return values["session"].(map[string]interface{}), nil
+		}
+		return values, nil
+	}
+	// We do want the return value wrapped in a session object, so wrap it if it
+	// isn't already.
+	if _, ok := values["session"]; !ok {
+		newmap := make(map[string]interface{})
+		newmap["session"] = values
+		return newmap, nil
+	}
+	return values, nil
+}
+
+// UserSessionsApp is an implementation of the App interface created to manage
+// user sessions.
+type UserSessionsApp struct {
+	db *sql.DB
+}
+
+// New returns a new *UserSessionsApp
+func New(db *sql.DB) *UserSessionsApp {
+	return &UserSessionsApp{
+		db: db,
+	}
+}
+
+// hasSessions returns whether or not the given user has a session already.
+func (u *UserSessionsApp) hasSessions(username string) (bool, error) {
+	query := `SELECT COUNT(s.*)
+              FROM user_sessions s,
+                   users u
+             WHERE s.user_id = u.id
+               AND u.username = $1`
```

Comment:
(another nice-to-have-but-happy-to-ignore)

this could be an EXISTS directly returning a boolean as well: `SELECT EXISTS(SELECT s.id FROM user_sessions s, users u WHERE s.user_id = u.id AND u.username = $1)`